cleverhans/future/torch/attacks/spsa.py

```diff
@@ -0,0 +1,187 @@
+    raise ValueError('number of inputs {} is different from number of labels {}'
+                     .format(len(x), len(y)))
+  if y is None:
+    y = torch.argmax(model_fn(x), dim=1)
+
+  # The rest of the function doesn't support batches of size greater than 1,
+  # so if the batch is bigger we split it up.
+  if len(x) != 1:
+    adv_x = []
+    for x_single, y_single in zip(x, y):
+      adv_x_single = spsa(model_fn=model_fn, x=x_single.unsqueeze(0), eps=eps,
+                          nb_iter=nb_iter, clip_min=clip_min, clip_max=clip_max,
+                          y=y_single.unsqueeze(0), targeted=targeted,
+                          early_stop_loss_threshold=early_stop_loss_threshold,
+                          learning_rate=learning_rate, delta=delta,
+                          spsa_samples=spsa_samples, spsa_iters=spsa_iters,
+                          is_debug=is_debug, sanity_checks=sanity_checks)
+      adv_x.append(adv_x_single)
+    return torch.cat(adv_x)
+
+  if eps < 0:
+    raise ValueError(
+        "eps must be greater than or equal to 0, got {} instead".format(eps))
+  if eps == 0:
+    return x
+
+  if clip_min is not None and clip_max is not None:
+    if clip_min > clip_max:
+      raise ValueError(
+          "clip_min must be less than or equal to clip_max, got clip_min={} and clip_max={}"
+          .format(clip_min, clip_max))
+
+  asserts = []
+
+  # If a data range was specified, check that the input was in that range
+  asserts.append(torch.all(x >= clip_min))
+  asserts.append(torch.all(x <= clip_max))
+
+  if is_debug:
+    print("Starting SPSA attack with eps = {}".format(eps))
+
+  perturbation = (torch.rand_like(x) * 2 - 1) * eps
+  _project_perturbation(perturbation, eps, x, clip_min, clip_max)
+  optimizer = optim.Adam([perturbation], lr=learning_rate)
+
+  for i in range(nb_iter):
+    def loss_fn(pert):
+      """
+      Margin logit loss, with correct sign for targeted vs untargeted loss.
+      """
+      logits = model_fn(x + pert)
+      loss_multiplier = 1 if targeted else -1
+      return loss_multiplier * _margin_logit_loss(logits, y.expand(len(pert)))
+
+    spsa_grad = _compute_spsa_gradient(loss_fn, x, delta=delta,
+                                       samples=spsa_samples, iters=spsa_iters)
+    perturbation.grad = spsa_grad
+    optimizer.step()
+
+    _project_perturbation(perturbation, eps, x, clip_min, clip_max)
+
+    loss = loss_fn(perturbation).item()
+    if is_debug:
+      print('Iteration {}: loss = {}'.format(i, loss))
+    if early_stop_loss_threshold is not None and loss < early_stop_loss_threshold:
+      break
+
+  adv_x = (x + perturbation).detach()
+
+  asserts.append(torch.all(torch.abs(adv_x - x) <= eps + 1e-6))
+  asserts.append(torch.all(adv_x >= clip_min))
+  asserts.append(torch.all(adv_x <= clip_max))
+
+  if sanity_checks:
+    assert np.all(asserts)
+
+  return adv_x
+
+
+def _project_perturbation(perturbation, epsilon, input_image, clip_min=-np.inf,
```

Comment:
This assumes the adversary uses an L_inf norm. Is it easy to generalize the implementation to L_inf and L2?

It would also make sense to reuse the function here to compute `clipped_perturbation` : https://github.com/tensorflow/cleverhans/blob/921cbdb73bb625c36bf749e4f2bda4f017ec8f09/cleverhans/future/torch/utils.py#L8

Reply:
The original implementation uses an L_inf threat model, but it should be easy to generalize to L2 as well. I'll work on it!

Reply:
Excellent, thanks! If this is simpler, we can merge the current PR and then make a follow-up adding L2 support. If that's better for you, it would be great to use the `clip_eta` util function and then I'll merge this PR. We can also wait for L2